platform/utm_mgrs_utils.cpp

```diff
@@ -0,0 +1,567 @@
+
+  if (rowInt == 'I' || (rowOrigin < 'I' && rowInt > 'I') || ((rowInt > 'I' || rowOrigin < 'I') && rollover)) {
+    rowInt++;
+  }
+
+  if (rowInt == 'O' || (rowOrigin < 'O' && rowInt > 'O') || ((rowInt > 'O' || rowOrigin < 'O') && rollover))
+  {
+    rowInt++;
+    if (rowInt == 'I')
+      rowInt++;
+  }
+
+  if (rowInt > 'V')
+    rowInt = rowInt - 'V' + 'A' - 1;
+
+  string twoLetter = {char(colInt), char(rowInt)};
+
+  return twoLetter;
+}
+
+int zone_to_100k(int i) {
+  int set = i % NUM_100K_SETS;
+  if (set == 0)
+    set = NUM_100K_SETS;
+  return set;
+}
+
+// Convert UTM parameters to lat,lon for WSG 84 ellipsoid.
+// If UTM parameters are valid lat and lon references are used to output calculated coordinates.
+// Otherwise function returns 'false'.
+bool UTMtoLatLon(double easting, double northing, int zone_number, char zone_letter, double &lat, double &lon)
+{
+  if (zone_number < 1 || zone_number > 60)
+    return false;
+
+  if (easting < 100000.0 || easting >= 1000000.0)
+    return false;
+
+  if (northing < 0.0 || northing > 10000000.0)
+    return false;
+
+  if (zone_letter<'C' || zone_letter>'X' || zone_letter == 'I' || zone_letter == 'O')
+    return false;
+
+  bool northern = (zone_letter >= 'N');
+  double x = easting - 500000.0;
+  double y = northing;
+
+  if (!northern)
+    y -= 10000000.0;
+
+  double m = y / K0;
+  double mu = m / (R * M1);
+
+  double p_rad = (mu +
+         P2 * sin(2.0 * mu) +
+         P3 * sin(4.0 * mu) +
+         P4 * sin(6.0 * mu) +
+         P5 * sin(8.0 * mu));
+
+  double p_sin = sin(p_rad);
+  double p_sin2 = p_sin * p_sin;
+
+  double p_cos = cos(p_rad);
+
+  double p_tan = p_sin / p_cos;
+  double p_tan2 = p_tan * p_tan;
+  double p_tan4 = p_tan2 * p_tan2;
+
+  double ep_sin = 1 - E * p_sin2;
+  double ep_sin_sqrt = sqrt(1 - E * p_sin2);
+
+  double n = R / ep_sin_sqrt;
+  double r = (1 - E) / ep_sin;
+
+  double c = E_P2 * p_cos * p_cos;
+  double c2 = c * c;
+
+  double d = x / (n * K0);
+  double d2 = d * d;
+  double d3 = d2 * d;
+  double d4 = d3 * d;
+  double d5 = d4 * d;
+  double d6 = d5 * d;
+
+  double latitude = (p_rad - (p_tan / r) *
+                    (d2 / 2.0 -
+                     d4 / 24.0 * (5.0 + 3.0 * p_tan2 + 10.0 * c - 4.0 * c2 - 9.0 * E_P2)) +
+                     d6 / 720.0 * (61.0 + 90.0 * p_tan2 + 298.0 * c + 45.0 * p_tan4 - 252.0 * E_P2 - 3.0 * c2));
+
+  double longitude = (d -
+                      d3 / 6.0 * (1.0 + 2.0 * p_tan2 + c) +
+                      d5 / 120.0 * (5.0 - 2.0 * c + 28.0 * p_tan2 - 3.0 * c2 + 8.0 * E_P2 + 24.0 * p_tan4)) / p_cos;
+
+  longitude = mod_angle(longitude + degToRad(zone_number_to_central_longitude(zone_number)));
+
+  lat = radToDeg(latitude);
+  lon = radToDeg(longitude);
+
+  return true;
+}
+
+// Convert MGRS parameters to UTM parameters and then use UTM to lat,lon conversion.
+bool MGRStoLatLon(double easting, double northing, int zone_code, char zone_letter, char square_code[2], double &lat, double &lon)
```

Comment:
ditto for return type?

Reply:
No more `bool` return type. This function now returns `optional<LatLon>`.